engine/context/src/main/java/io/deephaven/engine/context/QueryCompiler.java

```diff
@@ -683,27 +683,63 @@ private void maybeCreateClass(String className, String code, String packageName,
+
+    private static JavaCompiler getJavaCompiler() {
+        JavaCompiler localCompiler;
+        if ((localCompiler = JAVA_COMPILER) == null) {
+            synchronized (QueryCompiler.class) {
+                if ((localCompiler = JAVA_COMPILER) == null) {
+                    JAVA_COMPILER = localCompiler = ToolProvider.getSystemJavaCompiler();
+                }
+            }
+        }
+        if (localCompiler == null) {
+            throw new RuntimeException("No Java compiler provided - are you using a JRE instead of a JDK?");
+        }
+        return localCompiler;
+    }
+
+    /**
+     * While the JavaFileManager should be closed to clean up resources, using a singleton avoids repeated processing of
+     * the classpath, which is <b>very</b> expensive.
+     */
+    private static volatile JavaFileManager JAVA_FILE_MANAGER = null;
+
+    private static JavaFileManager getJavaFileManager() {
+        JavaFileManager localManager;
+        if ((localManager = JAVA_FILE_MANAGER) == null) {
+            synchronized (QueryCompiler.class) {
+                if ((localManager = JAVA_FILE_MANAGER) == null) {
+                    JAVA_FILE_MANAGER = localManager = getJavaCompiler().getStandardFileManager(null, null, null);
+                }
+            }
+        }
+        return localManager;
+    }
+
     private void maybeCreateClassHelper(String fqClassName, String finalCode, String[] splitPackageName,
             String rootPathAsString, String tempDirAsString) {
         final StringWriter compilerOutput = new StringWriter();
 
-        final JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
-        if (compiler == null) {
-            throw new RuntimeException("No Java compiler provided - are you using a JRE instead of a JDK?");
-        }
+        final JavaCompiler compiler = getJavaCompiler();
 
         final String classPathAsString = getClassPath() + File.pathSeparator + getJavaClassPath();
         final List<String> compilerOptions = Arrays.asList("-d", tempDirAsString, "-cp", classPathAsString);
 
-        final StandardJavaFileManager fileManager = compiler.getStandardFileManager(null, null, null);
+        final JavaFileManager fileManager = getJavaFileManager();
 
-        final boolean result = compiler.getTask(compilerOutput,
-                fileManager,
-                null,
-                compilerOptions,
-                null,
-                Collections.singletonList(new JavaSourceFromString(fqClassName, finalCode)))
-                .call();
+        final boolean result;
+        // the java file manager is not thread safe
+        synchronized (QueryCompiler.class) {
+            result = compiler.getTask(compilerOutput,
+                    fileManager,
+                    null,
+                    compilerOptions,
+                    null,
+                    Collections.singletonList(new JavaSourceFromString(fqClassName, finalCode)))
+                    .call();
+        }
```

Comment:
No way can we merge this unless you can convince me that compilation was already single-threaded somewhere under the hood (which I don't believe to be the case).

Reply:
I was going to make a similar comment.

Reply:
It's definitely not synchronized under the hood - go-client tests caught this via some concurrent modification exceptions.